### PR TITLE
Fixes the empty parameter cell and multi parameters on one line

### LINKED ISSUE #15169: "Run With Parameters" silently fails if parameters cell is empty/invalid
1. Create new notebook
2. Add parameter cell but leave it empty
3. Click `Run with Parameters`

Expected : Something to happen - either inform the user that there's no parameters to use or even just the run the notebook as is. 

Actual : Nothing, it silently fails.

The line here is currently broken in this scenario : https://github.com/microsoft/azuredatastudio/blob/main/src/sql/workbench/contrib/notebook/browser/notebookActions.ts#L315

When the cell is empty the cell source is an empty string which means the array after the split only has one element and thus that line throws an exception `TypeError: Cannot read property 'trim' of undefined`

Note that this will also happen for a cell with parameters if any of the lines don't contain an = (even if there are other valid lines)

### LINKED ISSUE #15171: "Run with Parameters" doesn't handle multiple parameters on the same line
1. Make a parameters cell with at least one line having multiple assignments on the same line : 

e.g. `y = 1; x = 2` (this is valid Python syntax)

1. Click `Run With Parameters`

Quick pick is shown like this 

image

It's unclear from the docs whether this is also something that Papermill will have issues with, but regardless we should definitely at least call it out in our own docs so people know that the syntax must be a specific format in order to work

### PR DESCRIPTION
**Pull request by VasuBhog**
This PR fixes #15169
This PR fixes #15171

If there is an empty parameter cell we will inform the user that there must be parameters in the parameter cell to run the action. Moreover, it catches when there is not a '=' in the parameter. Documentation will be more clear to users that they should put parameters within the parameter cell in order to utilize parameterization.

Also fixes the multiple parameters in one line by checking each parameter properly.

### CODE REVIEW

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 1]
```diff
@@ -310,9 +310,22 @@ export class RunParametersAction extends TooltipFromLabelAction {
 		let defaultParameters = new Map<string, string>();
 		editor.cells.forEach(cell => {
 			if (cell.isParameter) {
+				if (!cell.source) {
+					// If there is no parameters in the cell indicate to user to add them
+					this.notificationService.notify({
+						severity: Severity.Info,
+						message: localize('noParametersInCell', "This notebook cannot run with parameters until a there are parameters added to the parameter cell. [Learn more](https://docs.microsoft.com/sql/azure-data-studio/notebooks/notebooks-parameterization)."),
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:317]: Is this really true?  I would expect that we could still run - it just would run the Notebook as normal since there's no parameters to use.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 2]
```diff
@@ -310,9 +310,22 @@ export class RunParametersAction extends TooltipFromLabelAction {
 		let defaultParameters = new Map<string, string>();
 		editor.cells.forEach(cell => {
 			if (cell.isParameter) {
+				if (!cell.source) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:313]: Source can be an array of strings too - this doesn't seem to be handling the case where that's empty

Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:313]: What if the source is all whitespace or invalid parameters?

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 3]
```diff
@@ -310,9 +310,22 @@ export class RunParametersAction extends TooltipFromLabelAction {
...
+						severity: Severity.Info,
+						message: localize('noParametersInCell', "This notebook cannot run with parameters until a there are parameters added to the parameter cell. [Learn more](https://docs.microsoft.com/sql/azure-data-studio/notebooks/notebooks-parameterization)."),
+					});
+				}
 				for (let parameter of cell.source) {
-					let param = parameter.split('=', 2);
-					defaultParameters.set(param[0].trim(), param[1].trim());
+					if (parameter.includes('=')) {
+						// For syntax that includes multiple parameters on one line (ex. x=1;y=2)
+						let multiParams = parameter.split(';');
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:323]: And what if my parameter has a ; in it?

`var param1 = 'This will be parsed incorrectly; oh no!'

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts (hunk 1 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:317]: Yes, we should not run the normal notebook with no parameters as that would not be useful for parameterization. As such I think it would be better to tell the user to learn more about parameterization via the link and therefore help them understand that the parameter cell is for parameters.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts (hunk 2 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts]: Should be covered by checking the regex test for each line in the parameter cell.

Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts]: Should be covered by checking the regex test for each line in the parameter cell. Setting that emptyParameterCell to then only send the info message and not execute the notebook as well.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts (hunk 3 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:328]: discussed this will be documented that multi-line parameters are to not be typically used for parameters cell and therefore support for this would require a whole parsing/validation of the string which is something that will look into further for supporting multiple other languages for parameterization.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 4]
```diff
@@ -308,11 +308,31 @@ export class RunParametersAction extends TooltipFromLabelAction {
 		const editor = this._notebookService.findNotebookEditor(context);
 		// Set defaultParameters to the parameter values in parameter cell
 		let defaultParameters = new Map<string, string>();
+		let emptyParameterCell: boolean;
 		editor.cells.forEach(cell => {
 			if (cell.isParameter) {
+				// Check if cell is empty or contains just whitespace
+				for (let x of cell.source) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:315]: This doesn't work as expected if source is a string - it'll iterate over each character in the string and test that separately. So if that string has a space in it then it'll think it's empty.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 5]
```diff
@@ -308,11 +308,31 @@ export class RunParametersAction extends TooltipFromLabelAction {
...
+					emptyParameterCell = /^\s*$/.test(x);
+				}
+				if (!cell.source || emptyParameterCell) {
+					// If there is no parameters in the cell indicate to user to add them
+					this.notificationService.notify({
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:320]: Shouldn't you return after notifying this?

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 6]
```diff
@@ -308,11 +308,31 @@ export class RunParametersAction extends TooltipFromLabelAction {
 		const editor = this._notebookService.findNotebookEditor(context);
 		// Set defaultParameters to the parameter values in parameter cell
 		let defaultParameters = new Map<string, string>();
+		let emptyParameterCell: boolean;
 		editor.cells.forEach(cell => {
 			if (cell.isParameter) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:313]: Might want to consider breaking out after finding a parameter cell - and possibly display a warning (since having more than one parameter cell I would assume is not supported)

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 7]
```diff
@@ -308,11 +308,31 @@ export class RunParametersAction extends TooltipFromLabelAction {
...
+					emptyParameterCell = /^\s*$/.test(x);
+				}
+				if (!cell.source || emptyParameterCell) {
+					// If there is no parameters in the cell indicate to user to add them
+					this.notificationService.notify({
+						severity: Severity.Info,
+						message: localize('noParametersInCell', "This notebook cannot run with parameters until there are parameters added to the parameter cell. [Learn more](https://docs.microsoft.com/sql/azure-data-studio/notebooks/notebooks-parameterization)."),
+					});
+				}
 				for (let parameter of cell.source) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:325]: cell.source can be a single string - so again this logic isn't handling that correctly

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 8]
```diff
@@ -308,11 +308,31 @@ export class RunParametersAction extends TooltipFromLabelAction {
...
-					let param = parameter.split('=', 2);
-					defaultParameters.set(param[0].trim(), param[1].trim());
+					if (parameter.includes('=')) {
+						// For syntax that includes multiple parameters on one line (ex. x=1;y=2)
+						let multiParams = parameter.split(';');
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:328]: Given what we've discussed I think we should just remove this logic and tell users they need to have each parameter on its own line.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 9]
```diff
@@ -321,14 +341,14 @@ export class RunParametersAction extends TooltipFromLabelAction {
 		let inputParameters = new Map<string, string>();
 		let uriParams = new URLSearchParams();
 		// Store new parameter values to map based off defaultParameters
-		if (defaultParameters.size === 0) {
+		if (defaultParameters.size === 0 && !emptyParameterCell) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:344]: Why are we even getting down here if we have an empty parameters cell? As mentioned above I think we should just exit early in that case.

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts (hunk 6 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:313]: We have already ensured that users cant have more than two parameter cells in a notebook when they try to set another cell as a parameter cell we inform the user. I can add a return statement once we find one though so that we don't have to iterate through each cell.

Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:313]: A user can easily have a notebook they got elsewhere that doesn't adhere to this and we wouldn't detect that. You can address that in a separate PR if you wish but that's something I think we should address in general (possibly when opening a notebook as well)

Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:313]: Yeah, I can look into that further if a user tagged it from another notebook viewer. I think it's a general principle from Papermill as well to follow the parameters cell format. Any additional parameters would be tagged as "Injected-Parameters".

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts (hunk 9 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:339]: I do exit the notification and after seeing if a cell.parameter is found. However, this only exits the loop of the for loop and not the entire action and therefore still needs to make this condition to make sure we don't do anything further for an empty parameter cell.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts (hunk 7 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:327]: addressed this with new logic.

**Review by barbaravaldez — APPROVED** [review]

**Review by lucyzhang929 — APPROVED** [review]

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 10]
```diff
@@ -308,27 +308,42 @@ export class RunParametersAction extends TooltipFromLabelAction {
 		const editor = this._notebookService.findNotebookEditor(context);
 		// Set defaultParameters to the parameter values in parameter cell
 		let defaultParameters = new Map<string, string>();
+		let emptyParameterCell: boolean;
 		editor.cells.forEach(cell => {
 			if (cell.isParameter) {
+				// Check if cell is empty or contains just whitespace
+				const cellSource = typeof cell.source === 'string' ? [cell.source] : cell.source;
+				for (let x of cellSource) {
+					emptyParameterCell = /^\s*$/.test(x);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:317]: This doesn't seem correct - if the cell source is an array then you're just checking whether ANY of the values are empty. So if you have a bunch of non-empty values and then one empty value it'll think the cell is "empty", which isn't correct. 

Instead you should probably be doing `emptyParameterCell = cellSource.every(s => /^\s*$/.test(s))`

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 11]
```diff
@@ -308,27 +308,42 @@ export class RunParametersAction extends TooltipFromLabelAction {
...
+				if (emptyParameterCell) {
+					// If there is no parameters in the cell indicate to user to add them
+					this.notificationService.notify({
+						severity: Severity.Info,
+						message: localize('noParametersInCell', "This notebook cannot run with parameters until there are parameters added to the parameter cell. [Learn more](https://docs.microsoft.com/sql/azure-data-studio/notebooks/notebooks-parameterization)."),
+					});
+					return;
+				}
 				for (let parameter of cell.source) {
 					let param = parameter.split('=', 2);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:328]: You're assuming that the line is going to have a = in it. If it doesn't though it'll throw an error (null ref since param[1] is undefined)

If you're going to be assuming this then you should likely want to test that as well. You could add it to the empty-line check above if you wanted to consolidate the checks.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts (hunk 9 above)
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:339]: The return is in a forEach so it'll only exit that foreach.

Just convert the forEach function into an actual for loop instead and then you can return and not bother with these checks.

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]
You should also add tests - in this case primarily to demonstrate and enforce the syntax that you handle.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 12]
```diff
@@ -308,27 +308,42 @@ export class RunParametersAction extends TooltipFromLabelAction {
 		const editor = this._notebookService.findNotebookEditor(context);
 		// Set defaultParameters to the parameter values in parameter cell
 		let defaultParameters = new Map<string, string>();
+		let emptyParameterCell: boolean;
```
Comment (chlafreniere) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:311]: any reason why we can't just initialize this to false?

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 13]
```diff
@@ -308,27 +308,42 @@ export class RunParametersAction extends TooltipFromLabelAction {
...
 		editor.cells.forEach(cell => {
 			if (cell.isParameter) {
+				// Check if cell is empty or contains just whitespace
+				const cellSource = typeof cell.source === 'string' ? [cell.source] : cell.source;
```
Comment (chlafreniere) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:315]: We have similar logic in too many places. Is there anywhere where we're accessing the cell model from outside of that class where we want source as a string?

Not for this PR; but can you create a cleanup item? We probably have this logic in 10 different places. The cell model should either always return source as an array, or (if we need it), provide a way to return source as an array.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts (hunk 10 above)
Comment (chlafreniere) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:317]: Please add a comment here.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts (hunk 13 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:315]: As far as I know, I don't see references for wanting the source as a string. I can be sure to add this as a cleanup item / refactor.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts (hunk 12 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:311]: I can set it to false.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts (hunk 11 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts]: I think documentation will be clear to help users understand that '=' is needed for parameters, and only those values would be the ones able to be changed inside the parameter cell.

File: src/sql/workbench/contrib/notebook/test/browser/notebookActions.test.ts [hunk 14]
```diff
@@ -347,6 +347,51 @@ suite('Notebook Actions', function (): void {
...
+		let mockNotebookModel = new NotebookModelStub(undefined, undefined, testContents);
+
+		let action = new RunParametersAction('TestId', true, testUri, quickInputService, mockNotebookService.object, mockNotification.object);
+
+		mockNotebookEditor.setup(x => x.model).returns(() => mockNotebookModel);
+
+		// Run Parameters Action
+		await action.run(testUri);
+
+		assert.ok(actualMsg !== undefined, 'Should have received a notification');
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/test/browser/notebookActions.test.ts:390]: All of these checks are essentially checking the exact same thing so it seems redundant to have all 3. Just the last one is enough.

File: src/sql/workbench/contrib/notebook/test/browser/notebookActions.test.ts [hunk 15]
```diff
@@ -347,6 +347,51 @@ suite('Notebook Actions', function (): void {
...
+				kernelspec: {
+					name: 'python',
+					language: 'python',
+					display_name: 'Python 3'
+				}
+			},
+			nbformat: 4,
+			nbformat_minor: 5
+		};
+		let expectedMsg: string = localize('emptyParametersCell', "This notebook cannot run with parameters until there are parameters added to the parameter cell.[Learn more](https://docs.microsoft.com/sql/azure-data-studio/notebooks/notebooks-parameterization).");
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/test/browser/notebookActions.test.ts:368]: When making tests that verify strings it's best if you don't just copy the strings since then it makes it more annoying to have to remember to update this one as well if the actual string changes in the future. Just use the original string directly.

File: src/sql/workbench/contrib/notebook/test/browser/notebookActions.test.ts [hunk 16]
```diff
@@ -347,6 +347,51 @@ suite('Notebook Actions', function (): void {
 		assert.strictEqual(actualMsg, expectedMsg);
 	});
 
+	test('Run with Parameters Action with no empty parameter cell in notebook', async function (): Promise<void> {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/test/browser/notebookActions.test.ts:350]: This title doesn't seem to match what the test is actually doing

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 17]
```diff
@@ -308,14 +308,29 @@ export class RunParametersAction extends TooltipFromLabelAction {
 		const editor = this._notebookService.findNotebookEditor(context);
 		// Set defaultParameters to the parameter values in parameter cell
 		let defaultParameters = new Map<string, string>();
-		editor.cells.forEach(cell => {
+		let emptyParameterCell: boolean = false;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:311]: You don't need to define this up here anymore - just do

`const emptyParameterCell = cellSource.every(s => /^\s*$/.test(s));` below

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 18]
```diff
@@ -308,14 +308,29 @@ export class RunParametersAction extends TooltipFromLabelAction {
...
+					this.notificationService.notify({
+						severity: Severity.Info,
+						message: localize('noParametersInCell', "This notebook cannot run with parameters until there are parameters added to the parameter cell. [Learn more](https://docs.microsoft.com/sql/azure-data-studio/notebooks/notebooks-parameterization)."),
+					});
+					return;
+				}
 				for (let parameter of cell.source) {
-					let param = parameter.split('=', 2);
-					defaultParameters.set(param[0].trim(), param[1].trim());
+					if (parameter.includes('=')) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:327]: I would suggest adding a comment here explaining that we're purposely only handling very specific formatting for the parameters in case someone else has the same questions that I have had about all this stuff

**Review by Charles-Gagnon — APPROVED** [review]
Logic looks fine, but I would suggest adding more tests that cover all the different scenarios you expect (valid parameters, cell with empty string, cell with empty source array, etc). Tests should make sure to cover all expected branches of execution.
